include/h2o/http2_common.h

```diff
@@ -0,0 +1,294 @@
+    int exclusive;
+    uint32_t dependency;
+    uint16_t weight;
+} h2o_http2_priority_t;
+
+extern const h2o_http2_priority_t h2o_http2_default_priority;
+
+/* connection flow control window + alpha */
+#define H2O_HTTP2_DEFAULT_OUTBUF_SIZE 81920
+
+/* hpack */
+
+#define H2O_HTTP2_ENCODE_INT_MAX_LENGTH 5
+
+typedef struct st_h2o_hpack_header_table_t {
+    /* ring buffer */
+    struct st_h2o_hpack_header_table_entry_t *entries;
+    size_t num_entries, entry_capacity, entry_start_index;
+    /* size and capacities are 32+name_len+value_len (as defined by hpack spec.) */
+    size_t hpack_size;
+    size_t hpack_capacity;     /* the value set by SETTINGS_HEADER_TABLE_SIZE _and_ dynamic table size update */
+    size_t hpack_max_capacity; /* the value set by SETTINGS_HEADER_TABLE_SIZE */
+} h2o_hpack_header_table_t;
+
+typedef struct st_h2o_hpack_header_table_entry_t {
+    h2o_iovec_t *name;
+    h2o_iovec_t *value;
+    const char *err_desc; /* the recorded soft error description */
+} h2o_hpack_header_table_entry_t;
+
+void h2o_hpack_dispose_header_table(h2o_hpack_header_table_t *header_table);
+
+#define H2O_HPACK_PARSE_HEADERS_METHOD_EXISTS 1
+#define H2O_HPACK_PARSE_HEADERS_SCHEME_EXISTS 2
+#define H2O_HPACK_PARSE_HEADERS_PATH_EXISTS 4
+#define H2O_HPACK_PARSE_HEADERS_AUTHORITY_EXISTS 8
+
+int h2o_hpack_parse_headers(h2o_mem_pool_t *pool, const uint8_t *src, size_t len, h2o_hpack_header_table_t *header_table,
+                            const h2o_url_scheme_t **scheme, h2o_iovec_t *authority, h2o_iovec_t *method, h2o_iovec_t *path,
+                            h2o_headers_t *headers, int *pseudo_header_exists_map, size_t *content_length,
+                            h2o_cache_digests_t **digests, const char **err_desc);
+size_t h2o_hpack_encode_string(uint8_t *dst, const char *s, size_t len);
+void h2o_hpack_flatten_push_promise(h2o_buffer_t **buf, h2o_hpack_header_table_t *header_table, uint32_t stream_id,
+                                    size_t max_frame_size, const h2o_url_scheme_t *scheme, h2o_iovec_t authority,
+                                    h2o_iovec_t method, h2o_iovec_t path, h2o_headers_t *headers, uint32_t parent_stream_id);
+void h2o_hpack_flatten_response(h2o_buffer_t **buf, h2o_hpack_header_table_t *header_table, uint32_t stream_id,
+                                size_t max_frame_size, int status, h2o_headers_t headers, const h2o_iovec_t *server_name,
+                                size_t content_length);
+void h2o_hpack_flatten_request(h2o_buffer_t **buf, h2o_hpack_header_table_t *header_table, uint32_t stream_id,
+                               size_t max_frame_size, h2o_iovec_t method, h2o_url_t *url, h2o_headers_t *headers,
+                               int is_end_stream);
+void h2o_hpack_flatten_trailers(h2o_buffer_t **buf, h2o_hpack_header_table_t *header_table, uint32_t stream_id,
+                                size_t max_frame_size, h2o_header_t *headers, size_t num_headers);
```

Comment:
The "flatten" functions takes a immutable list of headers as the argument. Considering that, I think that the way of passing the headers should be by using a pointer to the list (i.e. array) and the number of items for maximum flexibility, instead of using `h2o_headers_t *` or `h2o_headers_t`.